### PR TITLE
[Doc] Set up the basis of Taichi specification

### PR DESCRIPTION
**Pull request by strongoier**
Related issue = #4602

This PR sets up the basic architecture of the specification, and includes the content of the preface, the "Introduction" section, and the "Values and types" subsection of the "Basic concepts" section.

### CODE REVIEW

File: docs/lang/articles/reference.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+sidebar_position: 9998
+---
+
+# Language Reference
+
+This article describes the syntax and semantics of the Taichi programming
+language.
+
+**To users**: If you have gone through user tutorials and still feel uncertain
+about your program behavior, then you are coming to the correct place. If you
```
Comment (k-ye) [inline docs/lang/articles/reference.md:11]:
```suggestion
about your program behavior, then you are in the right place. If you
```

File: docs/lang/articles/reference.md [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+the Taichi language and the Python language - code in the Taichi scope is
+treated as the former, while code in the Python scope is treated as the latter.
+It should be emphasized that this article is about *the Taichi language*.
+
+That said, because Taichi is embedded in Python, the syntax of Taichi is a
+subset of that of Python. To make life easier, this article is modeled after
+the [Python language reference](https://docs.python.org/3/reference/). The
+[notation](https://docs.python.org/3/reference/introduction.html#notation) and
+[lexical analysis](https://docs.python.org/3/reference/lexical_analysis.html)
+parts exactly follow Python, so familiarize yourself with them if you are new
```
Comment (k-ye) [inline docs/lang/articles/reference.md:35]: Let's tune down a bit..

```suggestion
parts exactly follow Python, please familiarize yourself with them if they seem new
```

File: docs/lang/articles/reference.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+to them.
+
+## Basic concepts
+
+Before detailing syntax and semantics in the next few chapters, many important
```
Comment (k-ye) [inline docs/lang/articles/reference.md:40]: many basic but important concepts and general evaluation principles specific to Taichi are listed here.

File: docs/lang/articles/reference.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+basic concepts and general evaluation principles unique to Taichi are
+introduced here.
+
+### Values and types
+
+Like other programming languages, each expression in Taichi will be evaluated
```
Comment (k-ye) [inline docs/lang/articles/reference.md:46]:
```suggestion
Like many other programming languages, each expression in Taichi will be evaluated
```

File: docs/lang/articles/reference.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+:::note
+For readers familiar with programming language terms, such behavior is inspired
+by [multi-stage programming](https://en.wikipedia.org/wiki/Multi-stage_programming)
+or [partial evaluation](https://en.wikipedia.org/wiki/Partial_evaluation).
+:::
+
+A Python value is simply a
+[Python object](https://docs.python.org/3/reference/datamodel.html),
+which directly comes from the following sources:
+- Free variables
```
Comment (k-ye) [inline docs/lang/articles/reference.md:62]: nit: I would put `literals` first because that is the simplest

File: docs/lang/articles/reference.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+- Arguments passed via `ti.template()`
+- Literals
+
+Furthermore, as long as all the operands of an operation are Python values,
+compile-time evaluation will take place and a result Python value will be
```
Comment (k-ye) [inline docs/lang/articles/reference.md:67]:
```suggestion
compile-time evaluation will take place, producing a result Python value.
```

This needs update. If I call `foo(x)`, where both `foo()` and `x` are Python values (i.e., `foo()` is not a Taichi function), all operands are Python value, but I still need to use it inside `ti.static`?

File: docs/lang/articles/reference.md [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+produced. For meta-programming purposes, Taichi provides an advanced
+environment for compile-time evaluation: `ti.static()`, where more operations
+are supported.
+
+A Python value only exists at compile time. After compile-time evaluation, all
```
Comment (k-ye) [inline docs/lang/articles/reference.md:72]: I feel like it's more like "transformed into Taichi values at JIT compile time"?

File: docs/lang/articles/reference.md [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+the remaining expressions will be evaluated to Taichi values at runtime.
+
+A Taichi value has a Taichi type, which is one of the following:
```
Comment (k-ye) [inline docs/lang/articles/reference.md:75]: I think we need to revamp our type system. At the end of the day, we can say:

which is one of the following:

* A primitive type
* A compound type
* An Ndarray
* ...

You can find more details in Type System.

WDYT?

File: docs/lang/articles/reference.md (hunk 6 above)
Comment (strongoier) [reply docs/lang/articles/reference.md:67]: No, you can directly use it.

File: docs/lang/articles/reference.md (hunk 7 above)
Comment (strongoier) [reply docs/lang/articles/reference.md:72]: I understand your point. The accurate way to say this should be "transformed into literals of Taichi AST". However, introducing Taichi AST here may complicate things even further. My purpose is to let users directly think about the final execution behavior based on their source code, and the actual Taichi value is indeed created at runtime.

File: docs/lang/articles/reference.md (hunk 8 above)
Comment (strongoier) [reply docs/lang/articles/reference.md:75]: I agree. They should be described in a unified way. My writing here is limited by the current situation :-(

**Review by k-ye — APPROVED** [review]
LGTM!

**Pull request merged**